index.d.ts

```diff
@@ -7,99 +7,240 @@
+   */
   addReference(
     xpath: string,
-    transforms?: ReadonlyArray<string>,
-    digestAlgorithm?: string,
+    transforms?: CanonicalizationAlgorithmType[],
+    digestAlgorithm?: HashAlgorithmType,
     uri?: string,
     digestValue?: string,
     inclusiveNamespacesPrefixList?: string,
     isEmptyUri?: boolean
   ): void;
-  checkSignature(xml: string): boolean;
+
+  /**
+   * Compute the signature of the given XML (using the already defined settings).
+   *
+   * @param xml The XML to compute the signature for.
+   * @returns If no callback is provided, returns `this` (the instance of SignedXml).
+   */
+  computeSignature(xml: string): SignedXml;
+
+  /**
+   * Compute the signature of the given XML (using the already defined settings).
+   *
+   * @param xml The XML to compute the signature for.
+   * @param callback A callback function to handle the signature computation asynchronously.
+   * @returns void
+   */
+  computeSignature(xml: string, callback: ComputeSignatureCallback): void;
+
+  /**
+   * Compute the signature of the given XML (using the already defined settings).
+   *
+   * @param xml The XML to compute the signature for.
+   * @param opts An object containing options for the signature computation.
+   * @returns If no callback is provided, returns `this` (the instance of SignedXml).
+   * @throws If the `location.action` option has an invalid action value.
+   */
+  computeSignature(xml: string, opts: ComputeSignatureOptions): SignedXml;
+
+  /**
+   * Compute the signature of the given XML (using the already defined settings).
+   *
+   * @param xml The XML to compute the signature for.
+   * @param opts An object containing options for the signature computation.
+   * @param callback A callback function to handle the signature computation asynchronously.
+   * @returns void
+   * @throws If the `location.action` option has an invalid action value.
+   */
   computeSignature(
     xml: string,
-    opts?: {
-      prefix?: string | undefined;
-      attrs?: { [key: string]: any } | undefined;
-      location?:
-        | {
-            reference: string;
-            action: "append" | "prepend" | "before" | "after";
-          }
-        | undefined;
-      existingPrefixes?: { [prefix: string]: string } | undefined;
-    }
+    opts: ComputeSignatureOptions,
+    callback: ComputeSignatureCallback
   ): void;
-  getOriginalXmlWithIds(): string;
+
+  /**
+   * Get the signature XML as a string.
+   *
+   * @returns The signature XML.
+   */
   getSignatureXml(): string;
+
+  /**
+   * Get the original XML with IDs as a string.
+   *
+   * @returns The original XML with IDs.
+   */
+  getOriginalXmlWithIds(): string;
+
+  /**
+   * Get the signed XML as a string.
+   *
+   * @returns The signed XML.
+   */
   getSignedXml(): string;
-  loadSignature(signatureNode: string | Node): void;
 }
 
 export interface KeyInfo {
   getKey(keyInfo?: Node[] | null): Buffer;
```

Comment:
Unless I'm missing something, this should actually be `getKey(keyInfo?:SelectedValue[] | undefined): string`. I say it returns `string` because `key` only ever is a `string`. I understand that `KeyLike` is accepted by the `verifySignature()` function, but we never use it. I suppose we could keep the broader type and then narrow it for specific implementations (`FileKeyInfo` and `StringKeyInfo`).

Reply:
It appears that `.getKey()` is called [here](https://github.com/node-saml/xml-crypto/blob/b80fe8c024bcc4ad39674264d156a4d98855c353/lib/signed-xml.js#L352) and that value is used in several places.

As for the `Buffer` situation, I believe that is probably a bug waiting to happen. We should make a PR to change that to read only `utf8`. In fact, some people complaining about keys not working over at `passport-saml` may be influenced by this and not even know it. Honestly, it escaped my attention. I can work on a PR for that change so that it doesn't mix in here.

Reply:
Actually, forget that... it can be a `Buffer` too...

Reply:
> It appears that `.getKey()` is called [here](https://github.com/node-saml/xml-crypto/blob/b80fe8c024bcc4ad39674264d156a4d98855c353/lib/signed-xml.js#L352) and that value is used in several places.

Yes but the parameter that is passed is not used at all so the return value is not affected by it. That is the issue I have with this function. It is even documented in the function that the parameter is not used.

Reply:
We don't need it, but apparently, others do use it:

- https://github.com/almerindo/RXML_Gen/blob/8865d62152bc97950e649c6f616998e218729a26/src/app/controllers/InvoiceController.js#L148
- https://github.com/aalonsog/test-cef-saml/blob/def29f60818fe252fd5df0850daf6ac859ce4133/saml-lib.js#L104

Some people apparently would like a different option:

 - https://github.com/teamdigitale/spid-passport/blob/5c0305293fdb0ffed63baff0ded563bfaaa7e007/index.js#L348

I'm open here. I'm ok with a breaking change here to clean things up, I'm also ok with renaming that argument to be `cert` so it is more correct and follows [this](https://github.com/node-saml/node-saml/pull/36) PR, which itself follows the specification [here](https://www.w3.org/TR/2008/REC-xmldsig-core-20080610/#sec-X509Data).

Actually, the more I think about it, the more I think we should just finish implementing the spec, thereby using the first argument, which should be called `cert`, not `key`. Thoughts @LoneRifle @ganesha289?

Reply:
I'm for adhering to spec. it's already problematic that we have many users who are unsure how xml-crypto should be correctly used. Spec adherence should help us avoid some of the confusion.

Reply:
Ok I understand now why people might need the certificate passed. I did not see that use-case. This is my keyinfo.

```
export class MyKeyInfo implements xmlcrypto.KeyInfo {

    private strID: string;
    private x509ID: string;
    private key: Buffer;
    public attrs: { [p: string]: any } = {};

    constructor(strID: string, x509ID: string, kiID: string, key: Buffer) {
        this.strID = strID;
        this.x509ID = x509ID;
        this.attrs.Id = kiID;
        this.key = key;
    }

    public getKey(): Buffer {
        return this.key;
    }

    public getKeyInfo(): string {
        let returnData = '';
        returnData += '<wsse:SecurityTokenReference ';
        returnData += 'xmlns:wsse ="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd" ';
        returnData += 'xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd" ';
        returnData += 'wsu:Id="' + this.strID + '">';
        returnData += '<wsse:Reference ';
        returnData += 'URI="#' + this.x509ID + '" ValueType="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-x509-token-profile-1.0#X509v3"/>';
        returnData += '</wsse:SecurityTokenReference>';

        return returnData;
    }
}
```

Reply:
> Unless I'm missing something, this should actually be `getKey(keyInfo?:SelectedValue[] | undefined): string`

Technically yes, that is the return type of `xpath.select`, but the way it is implemented now it will either return an array with one or more KeyInfo nodes, or an empty array if it is not found.

`this.keyInfo = xpath.select(".//*[local-name(.)='KeyInfo']", signatureNode);`

Reply:
> I still don't see why `keyInfo` is passed to `getKey()` though.

Maybe because of this use-case?

 - https://github.com/leandrob/saml20/blob/6763f8c7327205adfc1983bb4d1ba9d7c5bb1147/lib/validateSignature.js#L18
 - https://github.com/audibene-em/passport-azure-ad/blob/bb877b5527727249be3d4ada1809717092ac4c36/lib/passport-azure-ad/wsfedsaml.js#L74
 - https://github.com/lexplano/asic-verify/blob/dcc79ff71a23e6253dae91f22ff09f4999f254b6/lib/verifySignatures.js#L86

In fact, based on the common use-case, it seems that perhaps we shouldn't be defaulting to returning the `key` set in the `ctor`, but rather try to get the `key` from the `keyInfo` first and then return the `key` if that fails?